sdk/identity/azure-identity/src/samples/Azure-Cache-For-Redis/Jedis/Azure-AAD-Authentication-With-Jedis.md

```diff
@@ -127,21 +144,136 @@ while (i < maxTries) {
+// Note: Cache Host Name, Port, Username, Azure AD Access Token and ssl connections are required below.
+Jedis jedis = createJedisClient(cacheHostname, 6380, "<USERNAME>", accessToken, useSsl);
+
+int maxTries = 3;
+int i = 0;
+
+while (i < maxTries) {
+    try {
+        // Set a value against your key in the Redis cache.
+        jedis.set("Az:key", "testValue");
+        System.out.println(jedis.get("Az:key"));
+        break;
+    } catch (JedisException e) {
+        // Handle The Exception as required in your application.
+        e.printStackTrace();
+
+        // For Exceptions containing Invalid Username Password / Permissions not granted error messages, look at troubleshooting section at the end of document.
+
+        // Check if the client is broken, if it is then close and recreate it to create a new healthy connection.
+        if (jedis.isBroken()) {
+            jedis.close();
+            jedis = createJedisClient(cacheHostname, 6380, "<USERNAME>", tokenRefreshCache.getAccessToken(), useSsl);
+        }
+    }
+    i++;
+}
+// Close the Jedis Client
+jedis.close();
+
+// Helper Code
+private static Jedis createJedisClient(String cacheHostname, int port, String username, AccessToken accessToken, boolean useSsl) {
+    return new Jedis(cacheHostname, port, DefaultJedisClientConfig.builder()
+        .password(accessToken.getToken())
+        .user(username)
+        .ssl(useSsl)
+        .build());
+}
+
+/**
+ * The Token Cache to store and proactively refresh the Access Token.
+ */
+public static class TokenRefreshCache {
+    private final TokenCredential tokenCredential;
+    private final TokenRequestContext tokenRequestContext;
+    private final Timer timer;
+    private volatile AccessToken accessToken;
+    private final Duration refreshOffset;
+
+    /**
+     * Creates an instance of TokenRefreshCache
+     * @param tokenCredential the token credential to be used for authentication.
+     * @param tokenRequestContext the token request context to be used for authentication.
+     * @param refreshOffset the refresh offset to use to proactively fetch a new access token before expiry time.
+     */
+    public TokenRefreshCache(TokenCredential tokenCredential, TokenRequestContext tokenRequestContext, Duration refreshOffset) {
+        this.tokenCredential = tokenCredential;
+        this.tokenRequestContext = tokenRequestContext;
+        this.timer = new Timer();
+        this.refreshOffset = refreshOffset;
+    }
+
+    /**
+     * Gets the cached access token.
+     * @return the AccessToken
+     */
+    public AccessToken getAccessToken() {
+        if (accessToken != null) {
+            return  accessToken;
+        } else {
+            TokenRefreshTask tokenRefreshTask = new TokenRefreshTask();
+            accessToken = tokenCredential.getToken(tokenRequestContext).block();
+            timer.schedule(tokenRefreshTask, getTokenRefreshDelay());
+            return accessToken;
```

Comment:
Does something in here throw a reasonable exception or does this need to wrap and throw something meaningful?

Reply:
Timer, throws proper Exception types with cause as messages.